vulnfeeds/cmd/cvelist2osv/main.go

```diff
@@ -59,26 +65,419 @@ func extractConversionMetrics(cve cves.CVE5, refs []osvschema.Reference) {
+		for _, node := range c.Nodes {
+			if node.Operator != "OR" {
+				continue
+			}
+			for _, match := range node.CPEMatch {
+				if !match.Vulnerable {
+					continue
+				}
+				cpes = append(cpes, match.Criteria)
+				versionRange := osvschema.Range{}
+
+				// If no start version is given, assume the vulnerability starts from version "0".
+				if match.VersionStartIncluding == "" {
+					match.VersionStartIncluding = "0"
+				}
+
+				if match.VersionEndExcluding != "" {
+					addToVersionRange(&versionRange, match.VersionStartIncluding, "", match.VersionEndExcluding)
+				} else if match.VersionEndIncluding != "" {
+					addToVersionRange(&versionRange, match.VersionStartIncluding, match.VersionEndIncluding, "")
+				}
+				if len(versionRange.Events) != 0 {
+					versionRanges = append(versionRanges, versionRange)
+				}
+			}
+		}
+	}
+	if len(versionRanges) == 0 {
+		return nil, nil, fmt.Errorf("no versions extracted from CPEs")
+	}
+	return versionRanges, cpes, nil
 }
 
-// FromCVE5 creates a Vulnerability from a CVE5 object.
+// ExtractVersionsFromAffectedField extracts version ranges from a CVE 'affected' entry.
+// It handles various scenarios:
+// - If `defaultStatus` is "affected", it may calculate the inverse ranges (common for Linux).
+// - It iterates through `versions` entries with `status: "affected"`.
+// - It constructs ranges from `version`, `lessThan`, and `lessThanOrEqual` fields.
+// - For GitHub CVEs, it uses a special parser for version strings like "< 1.2.3".
+// - For git commits, it creates an introduced event.
+// - As a fallback, it may assume a single version means "fixed at this version, introduced at 0".
+//
+// Returns the extracted OSV ranges, the most frequent version type (e.g., "semver"), and any notes.
+func ExtractVersionsFromAffectedField(affected cves.Affected, cnaAssigner string) (versionRanges []osvschema.Range, rangeType string, notes []string) {
+
+	// Handle cases where a product is marked as "affected" by default, and specific versions are marked "unaffected".
+	if affected.DefaultStatus == "affected" {
+		// For Linux kernel CVEs, this logic is often handled by CPEs, so we skip it here.
+		if cnaAssigner == "Linux" {
+
+			notes = append(notes, "Skipping Linux Affected range versions in favour of CPE versions")
+			return nil, "", notes
+		}
+		// Calculate the affected ranges by finding the inverse of the unaffected ranges.
+		ranges, inverseNotes := findInverseAffectedRanges(affected)
+		notes = append(notes, inverseNotes...)
+
+		if len(ranges) != 0 {
+			return ranges, string(osvschema.RangeSemVer), notes
+		}
+		return nil, "", notes
+	}
+
+	versionTypesCount := make(map[string]int)
+
+	for _, vers := range affected.Versions {
+		if vers.Status != "affected" {
+			continue
+		}
+		if _, ok := versionTypesCount[vers.VersionType]; ok {
+			versionTypesCount[vers.VersionType]++
+		} else {
+			versionTypesCount[vers.VersionType] = 0
+		}
```

Comment:
I don't think this is needed, a map defaults to 0 as the empty value, so you should be able to just have the ++ line by itself.